packages/mobile/src/walletConnect/request.ts

```diff
@@ -29,14 +31,41 @@ export function* handleRequest({ method, params }: { method: string; params: any
   yield call(unlockAccount, account)
 
   switch (method) {
-    case SupportedActions.eth_signTransaction:
-      return (yield call(wallet.signTransaction.bind(wallet), params[0])) as EncodedTransaction
+    case SupportedActions.eth_signTransaction: {
+      // IMPORTANT: We need to normalize the transaction parameters
+      // WalletConnect v1 utils currently strips away important fields like `chainId`, `feeCurrency`, `gatewayFee` and `gatewayFeeRecipient`
+      // See https://github.com/WalletConnect/walletconnect-monorepo/blame/c6b26481c34848dbc9c49bb0d024bda907ec4599/packages/helpers/utils/src/ethereum.ts#L66-L86
+      // Also the dapp developer may have omitted some of the needed fields,
+      // so it's nice to be flexible and still allow the transaction to be signed (and sent) successfully
+
+      const kit: ContractKit = yield call(getContractKit)
+      const normalizer = new TxParamsNormalizer(kit.connection)
+      const rawTx = { ...params[0] }
+      // For now if `feeCurrency` is not set, we don't know whether it was stripped by WalletConnect v1 utils or intentionally left out
+      // to use CELO to pay for fees
+      if (!rawTx.feeCurrency) {
+        // This will use CELO to pay for fees if the user has a balance,
+        // otherwise it will fallback to the first currency with a balance
+        const feeCurrency: Currency = yield call(chooseFeeCurrency, Currency.Celo)
+        // Pass undefined to use CELO to pay for gas.
+        const feeCurrencyAddress: string | undefined =
+          feeCurrency === Currency.Celo ? undefined : yield call(getCurrencyAddress, feeCurrency)
+
+        rawTx.feeCurrency = feeCurrencyAddress
+        // Note: we're resetting gas and gasPrice here because if the feeCurrency has changed, we need to recalculate them
+        rawTx.gas = undefined
```

Comment:
One caveat here is that there are cases where gas estimation will fail even if the transaction will end up clearing. Clearing the gas may result in errors for the dApp. E.g. If a dApp developer is doing a two step approve and exchange and requesting both signatures together, they will set the gas on the second transaction because if estimateGas is run before the approve completes, execution will fail.

I saw in the Slack thread there is some mention of `estimateGas` being broken for non-CELO fee currencies, but I am not aware of any issue right now. It is true though that using a non-CELO fee currency uses more gas, by about 50k, so if the caller estimates using CELO and this code switches it to cUSD, the transaction may end up with too little gas.

Of these two scenarios, failures due to clearing the gas value seem more likely to me than failures due to not clearing it.

Reply:
Interesting problem, nice catch!

I think not having CELO to pay for gas is a somewhat common situation, especially for Valora users where they are encouraged to cash in cUSD/cEUR and not CELO. dApps are also usually not very good at catching it and show unhelpful errors when this happens.

My point is that the switch from CELO fee to cUSD/cEUR fee might happen somewhat often here. Don't really have the data to back it up, just my intuition (we could look it up).

How about just plain adding 50k to the `gas` field when this happens? Would that help solve for both use-cases or are there other complications I might be missing?

Reply:
> E.g. If a dApp developer is doing a two step approve and exchange and requesting both signatures together, they will set the gas on the second transaction because if estimateGas is run before the approve completes, execution will fail.

Do you mean that the dapp may call `eth_signTransaction` without having called estimateGas on their end and sent  individually?
So if it does that for approve and exchange, it will indeed fail for us.
Well, great catch!! 🥇 

So yeah ideally we should really not touch what the dapp is providing. But since we have this bug with WalletConnect v1 utils, we have to do something.
I really like what @gnardini is proposing then.

In addition, I'll also limit the workaround to when `type` is not set. Because in that case the TX data is sent as is. 
See https://github.com/WalletConnect/walletconnect-monorepo/blob/c6b26481c34848dbc9c49bb0d024bda907ec4599/packages/helpers/utils/src/ethereum.ts#L45

Reply:
Adding 50k to the gas value when setting a non-CELO fee currency seems reasonable. That should be the correct value as it is the value currently set in the `intrinsicGasForAlternativeFeeCurrency` on-chain 👌